server/src/main/java/org/apache/druid/client/CachingClusteredClient.java

```diff
@@ -598,7 +605,18 @@ private SortedMap<DruidServer, List<SegmentDescriptor>> groupSegmentsByServer(Se
     {
       final SortedMap<DruidServer, List<SegmentDescriptor>> serverSegments = new TreeMap<>();
       for (SegmentServerSelector segmentServer : segments) {
-        final QueryableDruidServer queryableDruidServer = segmentServer.getServer().pick(query);
+        Supplier<Set<String>> supplier = () -> {
+          boolean queryUnmanagedServers = query.context().getQueryUnmanagedServers();
+          if (queryUnmanagedServers) {
+            // If this is a test query, ignore source servers.
+            return dynamicConfigurationManager.getSourceClusterServers();
+          } else {
+            return dynamicConfigurationManager.getTargetCloneServers();
```

Comment:
Currently this is getting the list of historicals from na atom reference. Since this is the hot code, can we change the dynamic configuration manager to a read write lock and see the impact. I see you have a benchmark here as well. What are the results ?

Reply:
I think atomic reference should be faster:
https://stackoverflow.com/a/31578070

That said, for all the cases, the numbers are small enough (in 2-digit nanos) to not really matter.
I think the question becomes more relevant if we are doing some operations inside the lock.